### DISCUSSION TITLE
The issue of constraints on hanging nodes

### DISCUSSION THREAD

**Discussion by PEI0214** [libMesh/libmesh#4032] [General]
Hello, I would like to learn about how the constraints on 'hanging nodes' are handled in mesh adaptation. Where can I find relevant materials on this?

**Comment by jwpeterson** [comment]
It is discussed somewhat in the old 2006 libmesh paper, if you have access to that. Some of the information there may be a bit out of date, so your best bet is to look at the code itself, or ask (more specific) questions. Each of the FE classes implements a `compute_constraints()` API, so it might be instructive to look at one of those, e.g. `lagrange_compute_constraints()` in `src/fe/fe_lagrange.C`.

**Reply by PEI0214** [reply]
Hello, are the side midpoints and plane points treated the same way for hanging nodes?
![image](https://github.com/user-attachments/assets/00598564-97ee-4fb6-a315-562419781057)

**Reply by roystgnr** [reply]
They're calculated in the same code path, though the results differ: e.g. if those two hexes are Hex8s, the edge DoF will end up constrained by only its two adjacent nodes on the coarse edge, while the face DoF will end up constrained by all four coarse vertices on the coarse face.

There's also the complication that, whereas face DoFs are uniquely defined by the results of a single projection between coarse and fine face, edge DoFs in 3D are sometimes calculable from multiple coarse/fine face pairs sharing the edge; in those cases though we just use the results from the first calculation we reach while looping over elements.
